Mathlib/Tactic/Change.lean

```diff
@@ -0,0 +1,40 @@
+/-
+Copyright (c) 2023 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Tactic.Basic
+/-!
+
+# Tactic `change? _`
+
+This tactic is used to replace the goal with a definitionally equal one.
+The `_` is a general term and is intended to be used to guide which unfolding you want
+Lean to perform.
+-/
+
+/--  `change? term?` unifies the optional `term?` with the current goal, defaulting to the
+goal, if `term?` is not present.
+It then prints in the infoView the unified term.
+This is useful to replace the main goal after a `dsimp`.
+```lean
+example : (fun x : Nat => x) 0 = 1 := by
+  change? 0 = _  -- `change 0 = 1`
+```
+-/
+syntax (name := change?) "change?" (ppSpace colGt term)? : tactic
+
+open Lean Meta Elab.Tactic Std.Tactic.TryThis in
+elab_rules : tactic
+| `(tactic|change?%$tk $[$sop:term]?) => withMainContext do
+  let stx ← getRef
+  let expr ← match sop with
+    | none => getMainTarget
+    | some sop => do
+      let ex ← elabTerm sop none
+      let defeq? ← isDefEq ex (← getMainTarget)
```

Comment:
(Since we'd be using `getMainTarget` twice if using `elabTermEnsuringType`, I factored it out.)
```suggestion
      let tgt ← getMainTarget
      let ex ← withRef sop <| elabTermEnsuringType sop (← inferType tgt)
      let defeq? ← isDefEq ex tgt
```

Reply:
Ah, I was confused before about the defeq test, since I was thinking "these are all Props", but of course the defeq test is on the term, while the elabTerm is using `inferType`!  I'm normally slow, even in front of a computer, but on mobile, I cannot even test things out!  🙃